server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java

```diff
@@ -506,11 +523,76 @@ private List<UploadedIndexMetadata> writeIndexMetadataParallel(ClusterState clus
         return result;
     }
 
+    /**
+     * Invokes the index metadata upload listener.
+     */
+    private void invokeIndexMetadataUploadListeners(
+        List<IndexMetadata> newIndexMetadataList,
+        CountDownLatch latch,
+        List<Exception> exceptionList
+    ) {
+        for (IndexMetadataUploadListener listener : indexMetadataUploadListeners) {
+            // We are submitting the task for async execution to ensure that we are not blocking the cluster state upload
+            String listenerName = listener.getClass().getSimpleName();
+            String threadPoolName = listener.getThreadpoolName();
+            assert ThreadPool.THREAD_POOL_TYPES.containsKey(threadPoolName) && ThreadPool.Names.SAME.equals(threadPoolName) == false;
+            threadpool.executor(threadPoolName).execute(() -> {
+                try {
+                    listener.beforeNewIndexUpload(
+                        newIndexMetadataList,
+                        getIndexMetadataUploadActionListener(newIndexMetadataList, latch, exceptionList, listenerName)
+                    );
+                } catch (IOException e) {
```

Comment:
Do we also need to decrement latch on exception?

Reply:
Handling this in the `RemoteIndexPathUploader.java` class itself.